Field::Select to handle ActiveRecord enums correctly

pablobm opened the pull request:
Fixes https://github.com/thoughtbot/administrate/issues/1939

In the spirit of "make the change easy, then make the easy change", this PR includes a refactor of `Field::Select`, moving the logic out of the template and into the field class, where the actual functionality (and any future one) can be handled and tested more easily.

I have also written more exhaustive specs for the behaviour of this field, which helped me not breaking existing functionality while refactoring (or so I hope!).

littleforest reviewed (commented):
@pablobm these changes look good, but I'm not seeing that this PR solves the behavior noted in the ticket, mainly: 

> when I then update the resource and go back on the edit page the saved value is not the option selected

I tested in the UI with the Customer.kind. If I choose "vip" in the form and save, then "vip" is correctly getting saved as the customer kind, but when I go back to edit the record again, it shows "standard" as selected.

pablobm commented:
@littleforest - Oopsie, you are right. I had forgotten to remove the `collection` option from the Customers dashboard. I have just fixed it and added a feature spec. It should work now :crossed_fingers:

pablobm commented:
Hm. Something else has broken now. I see what it is, it's the product of a hack in a spec. I'm going to have lunch and hopefully solve it when I'm feeling less annoyed...

pablobm commented:
Fixed now! The required changes were:

- `spec/example_app/app/dashboards/customer_dashboard.rb` - Change the field declaration to not provide a `collection` and instead rely on the existing enum.
- `spec/features/form_spec.rb` - The existing examples relied on the field receiving options (via `#with_options`). This would turn the field into a `Field::Deferred` whose options can be modified. This is a bit of a hack... and so is my alternative, although I want to think that mine is less bad because it doesn't rely on private interfaces. Yeah, that's my excuse :stuck_out_tongue_closed_eyes: 

Ideally, fields should be proper object instead of being passed around as classes that are instantiated at some point. But that would require a larger refactor, so not for now.

littleforest reviewed (approved):
Looks great!
